DM-4848: Measure photometric repeatability and correctness of reported errors

wmwv opened the pull request:
Adds plots and calculation on photometric repeatability.  Calculate simple RMS and more formal SRD-motivated quantities.  Output to STDOUT and PNG plots in working directory.

jdswinbank commented on README.md:
````diff
@@ -8,17 +8,19 @@ Pre-requisites: install and declare the following
 
 ```
-eups distrib obs_decam obs_cfht
+eups distrib obs_decam obs_cfht validation_data_decam validation_data_cfht validate_drp
````
`eups distrib install obs_decam ...`.

jdswinbank commented on bin.src/validateCfht.py:
```diff
@@ -25,15 +25,12 @@
 import os.path
 import sys
 
-from lsst.validate.drp import checkAstrometry
```
`validateCfht.py` and `validateDecam.py` are very nearly the same thing modulo some numbers. Do they really need to be completely separate pieces of code? It would seem more natural to have a `ValidateTask` containing all the boilerplate and with subclasses for CFHT and DECam (and other instrumentation in future, one assumes), or even to have a single validation task with appropriate configuration parameters so it can address different instruments.

(I have effectively the same comments on both pieces of code, and won't repeat them.)

wmwv replied:
Yes!  Integrating and generalizing validateCfht.py and validateDecam.py and having instead a general script that reads a parameter file is DM-4901 and will tackled then.

jdswinbank commented on bin.src/validateCfht.py:
```diff
@@ -25,15 +25,12 @@
-
-    # Reference visit (the other viisits will be compared to this one
-    ref = 849375
+    visits = [849375, 850587]
```
This list of visits & CCDs seems to simply be a repetition of `examples/runCfht.list`. It would be better to define it once.

wmwv replied:
Will be done as part of DM-4901

jdswinbank commented on bin.src/validateCfht.py:
```diff
@@ -62,5 +58,5 @@ def defaultData(repo):
 
-    visitDataIds, refDataIds, good_mag_limit, medianRef, matchRef = defaultData(repo)
-    checkAstrometry.run(repo, visitDataIds, refDataIds, good_mag_limit, medianRef, matchRef)
+    args = defaultData(repo)
```
I realise it's only being use to copy from one line to the next, but `args` isn't a super descriptive name! It'd be better just to do away with it, I think

```
checkAstrometryPhotometry.run(repo, *defaultData(repo))
```

or similar.

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,67 @@
+import numpy as np
+import scipy.stats
+
+# To validate these width estimates, we can plot a single random realization (re-evaluate the cell or call the function again to get a new one).
```
Note the max. line length is 110 chars (throughout).

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,67 @@
+import scipy.stats
+
+# To validate these width estimates, we can plot a single random realization (re-evaluate the cell or call the function again to get a new one).
+def calcPA1(gv, magKey):
```
A docstring would help here in a few ways:
- While the cognoscenti will know where to look up what PA1 actually refers to, the casual reader won't -- please define it or provide a reference to an external definition;
- I can guess what `magKey` is, but on first read through of the code I have absolutely know idea what `gv` is supposed to be.
- There's a tuple with several fields being returned: please provide advice about how to interpret it.

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,67 @@
+    diffs = gv.aggregate(getRandomDiff, field=magKey)
+    means = gv.aggregate(np.mean, field=magKey)
+    rmsPA1, iqrPA1 = computeWidths(diffs)
+    return rmsPA1, iqrPA1, diffs, means
```
Why not return a struct or a named tuple?

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,67 @@
+
+
+def calcPA2(gv, magKey):
+    """Calculate PA2 values for min, design, stretch values of PF1.
```
While there is a docstring here, I have similar comments as on `calcPA1`, above.

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,67 @@
+# The SRD recommends computing repeatability from a histogram of magnitude differences for the same star measured on two visits (using a median over the diffs to reject outliers). Since our dataset includes N>2 measurements for each star, we select a random pair of visits for each star.
+# We also divide each difference by sqrt(2), because we want the RMS about the (unknown) mean magnitude, not the RMS difference, and convert from mags to mmag.
+# Note that this randomization still works for cases where we have only 2 obervations of the star.
+def getRandomDiff(array):
```
Given a function called `getRandomDiff` which takes an array as its argument, I might reasonably guess it returns the difference between two random elements of the array. This doesn't, and it's not immediately clear whether the comment above explains why or not.

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,67 @@
+    # not the most efficient way to extract a pair, but it's the easiest to write
+    copy = array.copy()
+    np.random.shuffle(copy)
+    return 1000*(copy[0] - copy[1])/(2**0.5)
```
Why `1000*`? Isn't `sqrt(2)` more readable than `2**0.5` (especially when it's the notation you use above)?

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,67 @@
+  
+    We estimate the width of the histogram in two ways: using a simple RMS, and using the interquartile range (scaled by the IQR/RMS ratio for a Gaussian). We do this for 50 different random realizations of the measurement pairs, to provide some estimate of the uncertainty on our RMS estimates due to the random shuffling (we could probably turn this into a more formal estimate somehow, but I'm no ...
+    While the SRD specifies that we should just compute the RMS directly, we haven't limited our sample to nonvariable stars as carefully as the SRD specifies, so using a more robust estimator like IQR will allow us to reject some outliers. It is also less sensitive some realistic sources of scatter the metric should include, however, such as bad zero points.
+    """
```
Where are the 50 pairs? After that docstring, I was expecting to see some sort of loop here, but I guess that comment is actually more relevant to some other part of the codebase. Could it migrate to where it's needed?

"I'm not going to bother at the moment" sounds like something that might be appropriate for an IPython notebook, but it grates in docstrings of reviewed code!

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,469 @@
+
+import lsst.afw.geom as afwGeom
+import lsst.afw.image as afwImage
+# import lsst.afw.coord as afwCoord
```
Please don't leave commented code in committed source.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,469 @@
+    dec = np.mean(cat.get('coord_dec'))
+    return ra, dec
+
+# Some thoughts from Paul Price on how to do the coordinate differences correctly:
```
So if this is correct, why aren't we doing it, rather than just including it as a comment?

wmwv replied:
Because it's not necessary for the photometric repeatability.  The goal is to do this correctly as part of calculating the astrometric SRD quantities, which is DM-4956.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,469 @@
+
+    @param[out]  pos_median -- median diff of positions in milliarcsec.  Float.
+
+    This is WRONG!
```
If it's wrong... why are we doing it? If it's just because it's good enough for now, can you file a ticket to indicate that it should be fixed in the future, rather than just leaving a warning in the code?

wmwv replied:
Thanks for catching this unneeded function.  This whole bit about the calculation of the position differences doesn't need to be in this ticket and should instead just be in DM-4956.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,469 @@
+
+    dataRefs = [dafPersist.ButlerDataRef(butler, vId) for vId in visitDataIds]
+
+    # retrieve the schema of the source catalog and extend it in order to add a field to record the ccd number
```
In fact, the line (171) which would extend the schema to include the CCD number is commented out.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,469 @@
+    # create the new extented source catalog
+    srcVis = SourceCatalog(newSchema)
+
+    for vId in visitDataIds:
```
I think here you're just annotating the source catalog with magnitudes calculated from the PSF flux. This seems like exactly what `TransformTask` is meant to do in a rather more generic way. `TransformTask` is unfortunately rather rather obscure (DM-4948), but it should be possible to reuse it here.

wmwv replied:
Yes, that's exactly right.

TransformTask sounds promising, but I fear that I can't quickly convert this to use TransformTask without some better documentation or example of its use.

Can you point me to what the equivalent way to write this would be using TransformTask?

Why is it a "Task"?  As far as I can tell, I get a source catalog back as the output of TransformTask.run().    I'm reading through

https://github.com/lsst/pipe_tasks/blob/5e752b3323af3387d8b07baf80128070c86db387/tests/testTransform.py#L169

and trying to figure out how this might work.

jdswinbank replied:
I'd be happy to sit down and go through this with you at some point (and augment the docs as we go), but I don't think I can commit to that on a sensible timescale for this review. For now this is fine.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,469 @@
+                        radius=matchRadius,
+                        RecordClass=SimpleRecord)
+
+    if False:
```
Again, please don't leave disabled code in the committed source unless there's a good reason (and that reason is well explained).

wmwv commented:
I've responded, added, squashed, and rebased.

I'd like to pursue the TransformTask further in a future ticket.
The generalization and consolidation of the DECam and CFHT code will occur on DM-4901.

jdswinbank commented:
Be aware that I've not finished reviewing -- I'm just distracted by a docs hackathon today. I'd like to get back to this tomorrow if you/it can wait -- is that ok?

wmwv commented:
Sure.  Every comment is useful.

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,276 @@
+import lsst.pipe.base as pipeBase
+
+def calcPA1(groupView, magKey):
+    """Calculate the photometric repeatability of measurements across a set of observations.
```
This docstring doesn't follow the standards as outlined at https://confluence.lsstcorp.org/display/LDMDG/Documentation+Standards#DocumentationStandards-FormattingoftheDocstringText. (Of course, I understand that those standards are up for debate and may be changed or replaced in the future.)

...that aside, it is a model of clarity & usefulness. Thanks!

(Same comments apply to several docstrings below, which I'm not going to call out on a case-by-case basis.)

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    >>> repoNameToPrefix("a/b/c")
+    "a_b_c"
+    >>> repoNameToPrefix("/bar/foo/")
+    "bar_foo"
```
I don't think this works as advertised. I get:

``` python
>>> repoNameToPrefix("/bar/foo/")
'bar_foo_'
```

It also seems that I can replicate all the examples (but correct) with a much simpler piece of code:

``` python
def repoNameToPrefix(repo):
    return repo.strip(os.sep).replace(os.sep, "_")
```

Is there a subtlety I'm missing?

wmwv replied:
Yes, thanks for catching that with repoNameToPrefix.  I just found that myself working on a different ticket yesterday, but forgot to backport.  I realize I hadn't thought completely through whether or not I wanted trailing underscores.

But your solution is much better.  Thanks!

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    mmatch = MultiMatch(newSchema, 
+                        dataIdFormat = {'visit': int, ccdKeyName: int},
+                        radius=matchRadius,
+                        RecordClass=SimpleRecord)
```
Should `RecordClass` be `SourceRecord`? What you have seems to work ok, in fairness, but `SourceRecord` seems correct according to the `MultiMatch` docs.

jdswinbank commented on python/lsst/validate/drp/calcSrd.py:
```diff
@@ -0,0 +1,276 @@
+    diffs = groupView.aggregate(getRandomDiffRmsInMas, field=magKey)
+    means = groupView.aggregate(np.mean, field=magKey)
+    rmsPA1, iqrPA1 = computeWidths(diffs)
+    return pipeBase.Struct(rms = rmsPA1, iqr = iqrPA1, 
```
Shouldn't have spaces around `=` here.

jdswinbank commented on python/lsst/validate/drp/plotAstrometryPhotometry.py:
```diff
@@ -0,0 +1,252 @@
+import scipy.stats
+from scipy.optimize import curve_fit
+
+from .calcSrd import calcPA1, calcPA2
```
`calcPA2` is unused.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    goodPsfMag = goodMatches.aggregate(np.mean, field=psfMagKey)  # mag
+    goodPsfMagRms = goodMatches.aggregate(np.std, field=psfMagKey)  # mag
+    goodPsfMagErr = goodMatches.aggregate(np.median, field=psfMagErrKey)
+    goodPsfMagNormDiff = goodMatches.aggregate(magNormDiff)
```
`goodPsfMagNormDiff` is unused. (Which is probably just as well, as it looks as though `magNormDiff` isn't returning anything -- see above.)

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    safeMatches = goodMatches.where(safeFilter)
+
+    # Pass field=psfMagKey so np.mean just gets that as its input
+    goodPsfMag = goodMatches.aggregate(np.mean, field=psfMagKey)  # mag
```
Not sure what the significance of the `# mag` is -- it seems clear this should be a magnitude without this.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    # Complete the match, returning a catalog that includes all matched sources with object IDs that can be used to group them.
+    matchCat = mmatch.finish()
+
+    # Create a mapping object that allows the matches to be manipulated as a mapping of object ID to catalog of sources.
```
And this one.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+        srcVis.extend(tmpCat, False)
+        mmatch.add(catalog=tmpCat, dataId=vId)
+
+    # Complete the match, returning a catalog that includes all matched sources with object IDs that can be used to group them.
```
Line too long.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    butler = dafPersist.Butler(repo)
+    dataset = 'src'
+
+    dataRefs = [dafPersist.ButlerDataRef(butler, vId) for vId in visitDataIds]
```
`dataRefs` is unused.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    magerr = cat.get('base_PsfFlux_magerr')
+    mag_avg = np.mean(mag)
+    N = len(mag)
+    normDiff = (mag - mag_avg) / magerr
```
No `return` -- this doesn't return anything.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    mag = cat.get('base_PsfFlux_mag')
+    magerr = cat.get('base_PsfFlux_magerr')
+    mag_avg = np.mean(mag)
+    N = len(mag)
```
`N` is unused.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+# import lsst.afw.coord as afwCoord
+
+from .plotAstrometryPhotometry import plotAstrometry, plotPhotometry, plotPA1
+from .calcSrd import computeWidths, getRandomDiff, calcPA1, calcPA2
```
`computeWidths` and `getRandomDiff` are unused.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+        Minimum average brightness (in magnitudes) for a star to be considered.
+    """
+
+    plotbase = repoNameToPrefix(repo)
```
I ran `python bin/validateDecam.py ./DECam/output/`. I got a bunch of plots saved with names like `._DECam_output_PA1.png` -- at first, I didn't even realise they existed. This probably isn't what you want.

Also, I think the variable name should be `plotBase` to follow standards.

wmwv replied:
Ah, good catch!  I didn't think about this.  I knew I wasn't completely happy with `repoNametoPrefix`, and still am not.  For now I'll explicitly handle this './' very common case.  I've added to doctest examples to define this.

```
    >>> repoNameToPrefix('./CFHT/output')
    'CFHT_output_'
    >>> repoNameToPrefix('.a/CFHT/output')
    'a_CFHT_output_'
```

Note this isn't quite consistent in some larger sense, but I'm fine with the code failing to produce distinct names for '.a/' and 'a/'.  The purpose was to generate different filenames for now.  Eventually this will all be hooked into the test harness as part of DM-2050.

jdswinbank commented on python/lsst/validate/drp/checkAstrometryPhotometry.py:
```diff
@@ -0,0 +1,514 @@
+    base = "_".join(dirnames[::-1])  # reverse to get order right
+    return base
+
+####
```
`####`?
